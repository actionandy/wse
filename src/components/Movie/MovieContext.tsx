import React, {createContext, useContext} from 'react';
import {useState} from 'react';
import {useCallback} from 'react';
import {MovieData} from './Movie';

interface MovieContextInterface {
  getMovieById: (id: string) => Promise<MovieData>;
  getAllMovies: (offset: number, limit: number) => Promise<MovieData[]>;
  getMovies: () => MovieData[];
};

const noop = () => {
  throw new Error('Movie context not provided.');
};

const MovieContext = createContext<MovieContextInterface>({getMovieById: noop, getAllMovies: noop, getMovies: noop});

export const MovieContextProvider: React.FC = ({children}) => {
  const [movies, setMovies] = useState<MovieData[]>([]);

  const getMovies: MovieContextInterface['getMovies'] = () => movies;

  const getAllMovies: MovieContextInterface['getAllMovies'] = useCallback(async (offset: number, limit: number) => {
    const fetchedMovies = await (await fetch(`/api/movies?offset=${offset}&limit=${limit + 1}`)).json();
    setMovies((movies) => {
      return [...movies, ...fetchedMovies].filter((movie: MovieData, i, all) => all.findIndex((_movie: MovieData) => _movie.ProviderTrackID === movie.ProviderTrackID) === i);
    });
    return fetchedMovies;
  }, []);

  const getMovieById: MovieContextInterface['getMovieById'] = useCallback(async (id: string) => {
    const movie = movies.find((movie, i) => movie.ProviderTrackID === id);
    if (movie) {
      return movie;
    } else {
      const fetchedMovie = await (await fetch(`/api/movies/${id}`)).json();
      setMovies((movies) => [...movies, fetchedMovie]);
      return fetchedMovie;
    }
  }, [movies]);

  return <MovieContext.Provider value={{getMovieById, getAllMovies, getMovies}}>{children}</MovieContext.Provider>;
};

export const useMovieContext = () => useContext(MovieContext);
