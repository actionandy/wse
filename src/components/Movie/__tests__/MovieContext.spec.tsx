import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react';
import {rest} from 'msw';
import {setupServer} from 'msw/node';
import mockedMovies from './__mocks__/movies.json';
import {MovieContextProvider, useMovieContext} from '../MovieContext';

const Dummy = () => {
  const res = useMovieContext();
  return (
    <>
      <div data-testid="movies">
        {res.getMovies().map((movie) => <span key={movie.ProviderTrackID}>{movie.ProviderTrackID}</span>)}
      </div>
      <button data-testid="loadMovies" onClick={() => res.getAllMovies(0, 36)}>Load movies</button>
    </>
  );
};

describe('MovieContext', () => {
  const server = setupServer();

  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it('LoadMoviesSuccess', async () => {
    server.use(
        rest.get(`/api/movies`, (request, response, context) => {
          return response(context.json(mockedMovies));
        }),
    );

    const {getByTestId, getByText} = render(
        <MovieContextProvider>
          <Dummy />
        </MovieContextProvider>,
    );

    fireEvent.click(getByTestId('loadMovies'));

    await waitFor(() => getByText('AOYloSgV5MA.P'));
    expect(getByTestId('movies')).toMatchSnapshot();
  });
});
