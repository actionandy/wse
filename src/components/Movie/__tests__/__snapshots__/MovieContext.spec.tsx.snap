// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContext LoadMoviesSuccess 1`] = `
<div
  data-testid="movies"
>
  <span>
    -9YZA_G7GJ0.P
  </span>
  <span>
    AOYloSgV5MA.P
  </span>
  <span>
    t1uJ0wSV2_U.P
  </span>
</div>
`;
